src/wh_server_dma.c

```diff
@@ -0,0 +1,307 @@
+#include "wolfhsm/wh_server.h"
+#include "wolfhsm/wh_error.h"
+#include "wolfhsm/wh_server_dma.h"
+
+/* TODO: if the Address allowlist ever gets large, we should consider a more
+ * efficient representation (requiring sorted array and binary search, or
+ * building a binary tree, etc.) */
+
+
+static int _checkAddrAgainstAllowList(const whServerDmaAddrList allowList, void* addr,
+                                      size_t size)
+{
+    uintptr_t startAddr = (uintptr_t)addr;
+    uintptr_t endAddr   = startAddr + size;
+
+    /* Check if the address range is fully within a allowlist entry */
+    for (int i = 0; i < WH_DMA_ADDR_ALLOWLIST_SIZE; i++) {
+        uintptr_t allowlistStartAddr = (uintptr_t)allowList[i].addr;
+        uintptr_t allowlistEndAddr   = allowlistStartAddr + allowList[i].size;
+
+        if (startAddr >= allowlistStartAddr && endAddr <= allowlistEndAddr) {
+            return WH_ERROR_OK;
+        }
+    }
+
+    return WH_ERROR_ACCESS;
+}
+
+static int _checkMemOperAgainstAllowList(const whServerDmaAddrAllowList* allowList,
+                                         whServerDmaOper oper, void* addr,
+                                         size_t size)
+{
+    int rc = WH_ERROR_OK;
+
+    /* If a read/write operation is requested, check the transformed address
+     * against the appropriate allowlist
+     *
+     * TODO: do we need to allowlist check on POST in case there are subsequent
+     * memory operations for some reason?
+     */
+    if (oper == WH_DMA_OPER_CLIENT_READ_PRE) {
+        rc = _checkAddrAgainstAllowList(allowList->readList, addr, size);
+    }
+    else if (oper == WH_DMA_OPER_CLIENT_WRITE_PRE) {
+        rc = _checkAddrAgainstAllowList(allowList->writeList, addr, size);
+    }
+
+    return rc;
+}
+
+int wh_Server_DmaRegisterCb32(whServerContext* server, whServerDmaClientMem32Cb cb)
+{
+    if (NULL == server || NULL == cb) {
+        return WH_ERROR_BADARGS;
+    }
+
+    server->dma.cb32 = cb;
+
+    return WH_ERROR_OK;
+}
+
+int wh_Server_DmaRegisterCb64(whServerContext* server, whServerDmaClientMem64Cb cb)
+{
+    if (NULL == server || NULL == cb) {
+        return WH_ERROR_BADARGS;
+    }
+
+    server->dma.cb64 = cb;
+
+    return WH_ERROR_OK;
+}
+
+int wh_Server_DmaRegisterAllowList(whServerContext*          server,
+                                   const whServerDmaAddrAllowList* allowlist)
+{
+    if (NULL == server || NULL == allowlist) {
```

Comment:
consider using allowlist == NULL to bypass address checking

Reply:
Fixed, and added test that NULL callbacks don't break things